evennia/contrib/utils/name_generator/namegen.py

```diff
@@ -0,0 +1,432 @@
+            "start": [],
+            "end": [],
+            "vowels": [ "a","a","a","a","a","e","i","i","i","y","u","o", ],
+            "length": (3,5),
+    },
+    "alien": {
+            "syllable": "C(C(V))(')(C)",
+            "consonants": [ 'q','q','x','z','v','w','k','h','b' ],
+            "start": ['x',],
+            "end": [],
+            "vowels": [ 'y','w','o','y' ],
+            "length": (1,5),
+    },
+
+}
+
+_RE_DOUBLES = re.compile(r'(\w)\1{2,}')
+
+# Load in optional settings
+
+custom_first_names = settings.NAMEGEN_FIRST_NAMES if hasattr(settings, "NAMEGEN_FIRST_NAMES") else []
+custom_last_names = settings.NAMEGEN_LAST_NAMES if hasattr(settings, "NAMEGEN_LAST_NAMES") else []
+
+if hasattr(settings, "NAMEGEN_FANTASY_RULES"):
+    _FANTASY_NAME_STRUCTURES |= settings.NAMEGEN_FANTASY_RULES
+
+if hasattr(settings, "NAMEGEN_REPLACE_LISTS") and settings.NAMEGEN_REPLACE_LISTS:
+    _FIRSTNAME_LIST = custom_first_names or _FIRSTNAME_LIST
+    _SURNAME_LIST = custom_last_names or _SURNAME_LIST
+
+else:
+    _FIRSTNAME_LIST += custom_first_names
+    _SURNAME_LIST += custom_last_names
+
+
+
+def fantasy_name(num=1, style="harsh", return_list=False):
+    """
+    Generate made-up names in one of a number of "styles".
+
+    Keyword args:
+        num (int)      - How many names to return.
+        style (string) - The "style" of name. This references an existing algorithm.
+        return_list (bool) - Whether to always return a list. `False` by default,
+                which returns a string if there is only one value and a list if more.
+    """
+    # validate num first
+    num = int(num)
+    if num < 1:
+        raise ValueError("Number of names to generate must be positive.")
+
+    if style not in _FANTASY_NAME_STRUCTURES:
+        raise ValueError(f"Invalid style name: '{style}'.")
+    style_dict = _FANTASY_NAME_STRUCTURES[style]
+
+    syllable = []
+    weight = 8
+    # parse out the syllable structure with weights
+    for key in style_dict["syllable"]:
+        # parentheses mean optional - allow nested parens
+        if key == "(":
+            weight = weight/2
+        elif key == ")":
+            weight = weight*2
+        else:
+            if key == "C":
+                type = "consonants"
+            elif key == "V":
+                type = "vowels"
+            else:
+                type = key
+            # append the sound type and weight
+            syllable.append( (type, int(weight)) )
+    
+    name_list = []
+    
+    # time to generate a name!
+    for n in range(num):
+        # build a list of syllables
+        length = random.randint(*style_dict['length'])
+        name = ""
+        for i in range(length):
+            # build the syllable itself
+            syll = ""
+            for sound, weight in syllable:
+                # random chance to skip this key; lower weights mean less likely
+                if random.randint(0,8) > weight:
```

Comment:
The weighting is a power of two determined by how deeply nested it is in parentheses, so the 8 is _technically_ arbitrary in the sense that I put a floor on how many nestings affect the weightings, i.e. 3 deep. Anything more deeply nested than that would then floor out at the 1 in 8.

I can make the floor depth configurable if you think it would be worthwhile, but I think anything that would go smaller would wind up being unnoticeable.

Reply:
Even is not allowing it to be tweaked (or not making much sense to tweak), it's still a good idea to move constants into the global context, at the top of the module (and to document them there) - an arbitrary number hidden in code like this means very little to the reader. At the very least, the number needs to be explained.